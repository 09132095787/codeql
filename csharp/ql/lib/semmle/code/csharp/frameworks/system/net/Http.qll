/** Provides definitions related to the `System.Net.Http` namespace. */

private import semmle.code.csharp.dataflow.ExternalFlow

/** Data flow for `System.Net.Http.HttpRequestOptions`. */
private class SystemNetHttpHttpRequestOptionsFlowModelCsv extends SummaryModelCsv {
  override predicate row(string row) {
    row =
      [
        "System.Net.Http;HttpRequestOptions;false;Add;(System.Collections.Generic.KeyValuePair<System.String,System.Object>);;Property[System.Collections.Generic.KeyValuePair<,>.Key] of Argument[0];Property[System.Collections.Generic.KeyValuePair<,>.Key] of Element of Argument[-1];value",
        "System.Net.Http;HttpRequestOptions;false;Add;(System.Collections.Generic.KeyValuePair<System.String,System.Object>);;Property[System.Collections.Generic.KeyValuePair<,>.Value] of Argument[0];Property[System.Collections.Generic.KeyValuePair<,>.Value] of Element of Argument[-1];value",
      ]
  }
}
