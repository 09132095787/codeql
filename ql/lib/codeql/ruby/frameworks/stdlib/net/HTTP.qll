private import codeql.ruby.AST
private import codeql.ruby.Concepts
private import codeql.ruby.dataflow.RemoteFlowSources
private import codeql.ruby.ApiGraphs
private import codeql.ruby.dataflow.internal.DataFlowPublic

/**
 * A shortcut for uses of Net::HTTP
 */
private API::Node netHTTP() { result = API::getTopLevelMember("Net").getMember("HTTP") }

/**
 * A call that returns the response body of a `Net::HTTP` request as a String.
 * ```ruby
 * req = Net::HTTP.new("example.com")
 * response = req.get("/")
 * body = response.body
 * ```
 */
private class NetHTTPRequestResponseBody extends CallNode {
  DataFlow::CallNode requestCall;

  NetHTTPRequestResponseBody() {
    exists(string methodName, API::Node requestCallNode |
      requestCall = requestCallNode.getAnImmediateUse()
    |
      // Net::HTTP.get(...)
      methodName = "get" and
      requestCallNode = netHTTP().getReturn(methodName) and
      this = requestCall
      or
      // Net::HTTP.post(...).body
      methodName in ["post", "post_form"] and
      requestCallNode = netHTTP().getReturn(methodName) and
      this = requestCallNode.getAMethodCall(["body", "read_body", "entity"])
      or
      // Net::HTTP.new(..).get(..).body
      methodName in [
          "get", "get2", "request_get", "head", "head2", "request_head", "delete", "put", "patch",
          "post", "post2", "request_post", "request"
        ] and
      requestCallNode = netHTTP().getInstance().getReturn(methodName) and
      this = requestCallNode.getAMethodCall(["body", "read_body", "entity"])
    )
  }

  /**
   * Gets the node representing the method call that initiates the request.
   * This may be different from the node which returns the response body.
   */
  DataFlow::Node getRequestCall() { result = requestCall }

  /**
   * Gets the node representing the URL of the request.
   * Currently unused, but may be useful in future, e.g. to filter out certain requests.
   */
  DataFlow::Node getURLArgument() { result = requestCall.getArgument(0) }
}

/**
 * A `Net::HTTP` call which initiates an HTTP request.
 */
class NetHTTPRequest extends HTTP::Client::Request::Range {
  private NetHTTPRequestResponseBody responseBody;

  NetHTTPRequest() { this = responseBody.getRequestCall().asExpr().getExpr() }

  override DataFlow::Node getResponseBody() { result = responseBody }

  override string getFramework() { result = "Net::HTTP" }
}
