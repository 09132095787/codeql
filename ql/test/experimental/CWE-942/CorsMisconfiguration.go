package main

import "net/http"

const (
	HeaderAllowOrigin      = "Access-Control-Allow-Origin"
	HeaderAllowCredentials = "Access-Control-Allow-Credentials"
)

func main() {
	{
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK: all origins are allowed,
			// but Access-Control-Allow-Credentials is not set.
			w.Header().Set("Access-Control-Allow-Origin", "*")
		})
	}
	{
		const Null = "null"
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// BAD: 'null' origin is allowed,
			// and Access-Control-Allow-Credentials is set to 'true'.
			w.Header().Set("Access-Control-Allow-Origin", "null")
			w.Header().Set("Access-Control-Allow-Credentials", "true")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// BAD: 'null' origin is allowed,
			// and `Access-Control-Allow-Credentials` is set to 'true':
			w.Header().Set(HeaderAllowOrigin, Null)
			w.Header().Set("Access-Control-Allow-Credentials", "true")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK: 'null' origin is allowed,
			// but Access-Control-Allow-Credentials is set to 'false'.
			w.Header().Set("Access-Control-Allow-Origin", Null)
			w.Header().Set("Access-Control-Allow-Credentials", "false")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK: 'null' origin is allowed,
			// but Access-Control-Allow-Credentials is not set.
			w.Header().Set("Access-Control-Allow-Origin", Null)
		})
	}

	{
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// BAD: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// and `Access-Control-Allow-Credentials` is set to 'true':
			origin := req.Header.Get("origin")
			w.Header().Set(HeaderAllowOrigin, origin)
			w.Header().Set("Access-Control-Allow-Credentials", "true")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// BAD: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// and `Access-Control-Allow-Credentials` is set to 'true':
			origin := req.Header.Get("origin")
			w.Header().Set("Access-Control-Allow-Origin", origin)
			w.Header().Set(HeaderAllowCredentials, "true")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// BAD: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// and `Access-Control-Allow-Credentials` is set to 'true':
			if origin := req.Header.Get("Origin"); origin != "" {
				w.Header().Set("Access-Control-Allow-Origin", origin)
			}
			w.Header().Set(HeaderAllowCredentials, "true")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK-ish: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// BUT `Access-Control-Allow-Credentials` is set to 'false':
			origin := req.Header.Get("origin")
			w.Header().Set("Access-Control-Allow-Origin", origin)
			w.Header().Set("Access-Control-Allow-Credentials", "false")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK-ish: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// BUT `Access-Control-Allow-Credentials` is not set:
			origin := req.Header.Get("origin")
			w.Header().Set("Access-Control-Allow-Origin", origin)
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK-ish: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// and `Access-Control-Allow-Credentials` is set to 'true',
			// BUT there is a check on the origin header:
			origin := req.Header.Get("origin")
			if allowedOrigin(origin) {
				w.Header().Set("Access-Control-Allow-Origin", origin)
			}
			w.Header().Set(HeaderAllowCredentials, "true")
		})
		allowedOrigins := map[string]bool{
			"example.com": true,
		}
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK-ish: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// and `Access-Control-Allow-Credentials` is set to 'true',
			// BUT there is a check on the origin header:
			origin := req.Header.Get("origin")
			isAllowed := allowedOrigins[origin]
			if isAllowed {
				w.Header().Set("Access-Control-Allow-Origin", origin)
			}
			w.Header().Set(HeaderAllowCredentials, "true")
		})
		http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
			// OK-ish: the `Access-Control-Allow-Origin` header is set using a user-defined value,
			// and `Access-Control-Allow-Credentials` is set to 'true',
			// BUT there is a check on the origin header:
			origin := req.Header.Get("origin")
			allowAllHosts := true
			allowedHost := isAllowedHost([]string{"example.com"}, origin)
			if allowAllHosts {
				w.Header().Set("Access-Control-Allow-Origin", "*")
			} else if allowedHost {
				w.Header().Set("Access-Control-Allow-Origin", origin)
				w.Header().Set(HeaderAllowCredentials, "true")
			}
		})
	}
}

func isAllowedHost(allowed []string, try string) bool {
	for _, v := range allowed {
		if try == v {
			return true
		}
	}
	return false
}

func allowedOrigin(origin string) bool {
	if origin == "example.com" {
		return true
	}
	return false
}
