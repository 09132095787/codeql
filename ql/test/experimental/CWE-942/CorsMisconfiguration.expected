| CorsMisconfiguration.go:15:4:15:53 | call to Set | access-control-allow-origin header is set to `*`, and access-control-allow-credentials is set to `true` |
| CorsMisconfiguration.go:21:4:21:41 | call to Set | access-control-allow-origin header is set to `*`, and access-control-allow-credentials is set to `true` |
| CorsMisconfiguration.go:41:4:41:56 | call to Set | access-control-allow-origin header is set to `null`, and access-control-allow-credentials is set to `true` |
| CorsMisconfiguration.go:47:4:47:42 | call to Set | access-control-allow-origin header is set to `null`, and access-control-allow-credentials is set to `true` |
| CorsMisconfiguration.go:68:4:68:53 | call to Set | access-control-allow-origin header is set to `*`, and access-control-allow-credentials is set to `true` |
| CorsMisconfiguration.go:86:4:86:44 | call to Set | access-control-allow-origin header is set to a user-defined value, and access-control-allow-credentials is set to `true` |
| CorsMisconfiguration.go:93:4:93:56 | call to Set | access-control-allow-origin header is set to a user-defined value, and access-control-allow-credentials is set to `true` |
| CorsMisconfiguration.go:100:5:100:57 | call to Set | access-control-allow-origin header is set to a user-defined value, and access-control-allow-credentials is set to `true` |
